snowflakeid/src/main/java/snowflakeid/SnowflakeId.java

```diff
@@ -0,0 +1,114 @@
+package snowflakeid;
+
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.security.MessageDigest;
+import java.time.Clock;
+import java.time.Instant;
+import java.time.ZoneId;
+import java.util.Random;
+import java.util.concurrent.atomic.AtomicLong;
+
+public class SnowflakeId {
+    private static volatile SnowflakeId instance;
+
+    private static final long defaultEpoch = 1759536000000L;
+    private static final long machineIdBits = 10L;
+    private static final long sequenceBits = 12L;
+
+    private static final long maxMachineId = ~(-1L << machineIdBits);
+
+    private long machineId = getMachineId();
+    private static final long machineIdShift = sequenceBits;
+    private static final long timestampLeftShift = sequenceBits + machineIdBits;
+    private static final long sequenceMask = ~(-1L << sequenceBits);
+
+    private Clock clock = Clock.systemUTC();
+
+    private volatile long lastTimestamp = -1L;
+    private long epoch;
+    private AtomicLong sequence = new AtomicLong(0);
+
+    private SnowflakeId(long epoch) {
+        this.epoch = epoch;
+    }
+
+    public static SnowflakeId getInstance(Long epoch) {
+        if (instance == null) {
+            synchronized (SnowflakeId.class) {
+                if (instance == null) {
+                    long finalEpoch = epoch == null ? defaultEpoch : epoch;
+                    instance = new SnowflakeId(finalEpoch);
+                }
+            }
+        }
+        return instance;
+    }
+
+    public long nextId() {
+        long timestamp = timeGen();
+
+        if (timestamp < lastTimestamp) {
+            throw new RuntimeException(String.format("Clock moved backwards. Refusing to generate ID for %d milliseconds", lastTimestamp - timestamp));
+        }
+
+        if (lastTimestamp == timestamp) {
+            sequence.set((sequence.get() + 1) & sequenceMask);
+            if (sequence.get() == 0) {
+                timestamp = tilNextMillis(lastTimestamp);
+            }
+        } else {
+            sequence.set(0);
+        }
+
+        lastTimestamp = timestamp;
+
+        return ((timestamp - epoch) << timestampLeftShift) |
+                (machineId << machineIdShift) |
+                sequence.get();
+    }
+
+    private long tilNextMillis(long lastTimestamp) {
+        long timestamp = timeGen();
+        while (timestamp <= lastTimestamp) {
```

Comment:
This will loop for a very very very long time in a heavy multi-thread environment.